File: BE/DatnBackend.Api/Models/QaPost.cs
namespace DatnBackend.Api.Models;

public class QaPost
{
    public string Id { get; set; } = "";
    public string UserId { get; set; } = "";
    public string UserName { get; set; } = "";
    public string UserAvatar { get; set; } = "";
    public string Title { get; set; } = "";
    public string Content { get; set; } = "";
    public string? LessonId { get; set; } // optional, null if general
    public List<string> Tags { get; set; } = new();
    public int AnswerCount { get; set; }
    public int UpvoteCount { get; set; }
    public bool IsSolved { get; set; }
    public DateTime CreatedAt { get; set; }
}

public class QaPostDto
{
    public string Id { get; init; } = "";
    public string UserId { get; init; } = "";
    public string UserName { get; init; } = "";
    public string UserAvatar { get; init; } = "";
    public string Title { get; init; } = "";
    public string Content { get; init; } = "";
    public string? LessonId { get; init; }
    public List<string> Tags { get; init; } = new();
    public int AnswerCount { get; init; }
    public int UpvoteCount { get; init; }
    public bool IsSolved { get; init; }
    public DateTime CreatedAt { get; init; }
    /// <summary>True nếu bài này chưa được user xem (tạo sau LastSeenQaAt và không phải của user)</summary>
    public bool IsNew { get; init; }

    public QaPostDto(QaPost post, string? viewerUserId, DateTime? lastSeenAt)
    {
        Id = post.Id; UserId = post.UserId; UserName = post.UserName;
        UserAvatar = post.UserAvatar; Title = post.Title; Content = post.Content;
        LessonId = post.LessonId; Tags = post.Tags; AnswerCount = post.AnswerCount;
        UpvoteCount = post.UpvoteCount; IsSolved = post.IsSolved; CreatedAt = post.CreatedAt;
        IsNew = viewerUserId != null
            && post.UserId != viewerUserId
            && (lastSeenAt == null || post.CreatedAt > lastSeenAt);
    }
}

File: mobile/lib/models/qa_post.dart
class QaPost {
  final String id;
  final String title;
  final String content;
  final String authorId;
  final String authorName;
  final String authorAvatar;
  final List<String> tags;
  final String? lessonId;
  final int answerCount;
  final int upvotes;
  final bool isSolved;
  final DateTime createdAt;
  final bool isNew;

  const QaPost({
    required this.id,
    required this.title,
    required this.content,
    required this.authorId,
    required this.authorName,
    required this.authorAvatar,
    required this.tags,
    this.lessonId,
    required this.answerCount,
    required this.upvotes,
    required this.isSolved,
    required this.createdAt,
    this.isNew = false,
  });

  factory QaPost.fromJson(Map<String, dynamic> json) {
    final tagList = (json['tags'] as List<dynamic>? ?? [])
        .map((e) => e.toString())
        .toList();

    final author = json['author'] as Map<String, dynamic>?;

    return QaPost(
      id: json['_id'] as String? ?? json['id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      content: json['content'] as String? ?? '',
      authorId: author?['_id'] as String? ?? json['authorId'] as String? ?? json['userId'] as String? ?? '',
      authorName: author?['name'] as String?
          ?? json['authorName'] as String?
          ?? json['userName'] as String?
          ?? '',
      authorAvatar: author?['avatar'] as String?
          ?? json['authorAvatar'] as String?
          ?? json['userAvatar'] as String?
          ?? '',
      tags: tagList,
      lessonId: json['lessonId'] as String?,
      answerCount: json['answerCount'] as int? ?? 0,
      upvotes: json['upvoteCount'] as int? ?? json['upvotes'] as int? ?? 0,
      isSolved: json['isSolved'] as bool? ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
      isNew: json['isNew'] as bool? ?? false,
    );
  }
}
